perf/perf.md

```diff
@@ -0,0 +1,121 @@
+# Perf <!-- omit in toc -->
+
+| Lifecycle Stage | Maturity      | Status | Latest Revision |
+| --------------- | ------------- | ------ | --------------- |
+| 1A              | Working Draft | Active | r0, 2022-11-16  |
+
+Authors: [@marcopolo]
+
+Interest Group: [@marcopolo], [@mxinden], [@marten-seemann]
+
+[@marcopolo]: https://github.com/mxinden
+[@mxinden]: https://github.com/mxinden
+[@marten-seemann]: https://github.com/marten-seemann
+
+# Table of Contents <!-- omit in toc -->
+- [Context](#context)
+- [Protocol](#protocol)
+- [Benchmarks](#benchmarks)
+  - [Single connection throughput](#single-connection-throughput)
+  - [Handshakes per second](#handshakes-per-second)
+- [Security Considerations](#security-considerations)
+- [Prior Art](#prior-art)
+
+# Context
+
+The `perf` protocol represents a standard benchmarking protocol that we can use
+to talk about performance within and across libp2p implementations. This lets us
+analyze peformance, guide improvements, and protect against regressions.
+
+# Protocol
+
+The `/perf/1.0.0` protocol (from here on referred to as simply `perf`) is a
+client driven set of benchmarks. To not reinvent the wheel, this perf protocol
+is almost identical to Nick Bank's [_QUIC Performance_
+Internet-Draft](https://datatracker.ietf.org/doc/html/draft-banks-quic-performance#section-2.3)
+but adapted to libp2p.
+
+
+The protocol is as a follows:
+
+Client:
+
+1. Open a libp2p stream to the server.
+2. Tell the server how many bytes we want the server to send us as a single
+   big-endian uint64 number. Zero is a valid number, so is the max uint64 value.
+3. Write some amount of data to the stream.
```

Comment:
Why? Sending data isn't very useful without a response from the server.

Consider making this a new upload handler, where the server explicitly acknowledges that it received all the sent data.

Reply:
Would you like to see an explicit acknowledgement for correctness, i.e. to make sure the data really made it from A to B?

I don't think we need this correctness guarantee. In my eyes this is a performance testing protocol only.

If the client wants a response from the server, it can request a single byte by sending a `1` early on.

Reply:
I can achieve amazing performance if I just disable my congestion controller and dump all packets into the network. Sure, most of them will be lost due to congestion, but since the sending and receiving part are completely decoupled in this protocol, I wouldn't even notice.

I don't understand what the point of the client sending data after the initial integer in this protocol is.

Reply:
The networking stack should guarantee that the bytes made it to the other end. We aren't checking for correctness here, we assume the networking stack works. If we read an EOF on the server and no other error while reading from the stream we assume we got all the bytes intended for us.

> I can achieve amazing performance if I just disable my congestion controller and dump all packets into the network

Your networking stack should retry the lost packets and the server doesn't progress until it sees the write side is closed. So if you had reliable delivery, you wouldn't get amazing performance this way.

That said, we could add some protection here where the server sends back a uint64 of bytes received as the first chunk of data when it starts sending the requested response data. The client can then check this number.

Reply:
I still don’t see why the client would send anything after the integer. The server doesn’t act upon this data, it’s just added complexity of draining the stream.

Reply:
This is how you benchmark the client's upload speed.

Reply:
That doesn’t work, as the server doesn’t send any response once it receives the EOF. I’d be in favor of cleanly separating uploads and downloads into two separate handlers.